cli/src/cmd/forge/script/broadcast.rs

```diff
@@ -216,135 +214,269 @@ impl ScriptArgs {
-    /// and/or gas calculations).
-    async fn handle_chain_requirements(
+    /// Takes the collected transactions and executes them locally before converting them to
+    /// [`TransactionWithMetadata`] with the appropriate gas execution estimation. If
+    /// `--skip-simulation` is passed, then it will skip the execution.
+    async fn fills_transactions_with_gas(
         &self,
-        txes: VecDeque<TransactionWithMetadata>,
-        provider: Arc<RetryProvider>,
-        chain: u64,
+        txs: BroadcastableTransactions,
+        script_config: &mut ScriptConfig,
+        decoder: &mut CallTraceDecoder,
+        known_contracts: &ContractsByArtifact,
     ) -> eyre::Result<VecDeque<TransactionWithMetadata>> {
-        let mut is_legacy = self.legacy;
-        if let Chain::Named(chain) = Chain::from(chain) {
-            is_legacy |= chain.is_legacy();
+        let gas_filled_txs = if self.skip_simulation {
+            println!("\nSKIPPING ON CHAIN SIMULATION.");
+            txs.into_iter()
+                .map(|btx| {
+                    let mut tx = TransactionWithMetadata::from_typed_transaction(btx.transaction);
+                    tx.rpc = btx.rpc;
+                    tx
+                })
+                .collect()
+        } else {
+            self.onchain_simulation(
+                txs,
+                script_config,
+                decoder,
+                known_contracts,
+            )
+            .await
+            .map_err(|err| {
+                eyre::eyre!(
+                    "{err}\n\nTransaction failed when running the on-chain simulation. Check the trace above for more information."
+                )
+            })?
         };
+        Ok(gas_filled_txs)
+    }
+
+    /// Returns all transactions of the [`TransactionWithMetadata`] type in a list of
+    /// [`ScriptSequence`]. List length will be higher than 1, if we're dealing with a multi
+    /// chain deployment.
+    ///
+    /// Each transaction will be added with the correct transaction type and gas estimation.
+    async fn bundle_transactions(
+        &self,
+        transactions: VecDeque<TransactionWithMetadata>,
+        target: &ArtifactId,
+        config: &mut Config,
+        returns: HashMap<String, NestedValue>,
+    ) -> eyre::Result<Vec<ScriptSequence>> {
+        // User might be using both "in-code" forks and `--fork-url`.
+        let last_rpc = &transactions.back().expect("exists; qed").rpc;
+        let is_multi_deployment = transactions.iter().any(|tx| &tx.rpc != last_rpc);
+
+        let mut total_gas_per_rpc: HashMap<RpcUrl, U256> = HashMap::new();
+
+        // Batches sequence of transactions from different rpcs.
+        let mut new_sequence = VecDeque::new();
+        let mut manager = ProvidersManager::default();
+        let mut deployments = vec![];
+
+        // Peeks next transaction to figure out if it's the same rpc as the current batch.
+        let mut txes_iter = transactions.into_iter().peekable();
```

Comment:
peekable is unused here?

Reply:
it's used to check if the next rpc url is different. If so, it creates a `ScriptSequence` from all the collected transactions so far.